src/lib/server/ring_buffer.hpp

```diff
@@ -0,0 +1,182 @@
+  RingBufferIterator _start_position = RingBufferIterator(_data);
+  RingBufferIterator _current_position = RingBufferIterator(_data);
+};
+
+// Dedicated buffer for read operations. The ring buffer gets extended by methods for reading different data types.
+template <typename SocketType>
+class ReadBuffer : public RingBuffer {
+ public:
+  explicit ReadBuffer(std::shared_ptr<SocketType> socket) : _socket(socket) {}
+
+  // Extract numerical values from buffer. Values will be converted into the correct byte order if necessary.
+  template <typename T, typename std::enable_if_t<std::is_same_v<uint16_t, T> || std::is_same_v<int16_t, T>, int> = 0>
+  T get_value() {
+    _receive_if_necessary(sizeof(T));
+    T network_value = 0;
+    std::copy_n(_start_position, sizeof(T), reinterpret_cast<char*>(&network_value));
+    _start_position += sizeof(T);
+    return ntohs(network_value);
+  }
+
+  template <typename T, typename std::enable_if_t<std::is_same_v<uint32_t, T> || std::is_same_v<int32_t, T>, int> = 0>
+  T get_value() {
+    _receive_if_necessary(sizeof(T));
+    T network_value = 0;
+    std::copy_n(_start_position, sizeof(T), reinterpret_cast<char*>(&network_value));
+    _start_position += sizeof(T);
+    return ntohl(network_value);
+  }
+
+  template <typename T, typename std::enable_if_t<!std::is_same_v<uint16_t, T> && !std::is_same_v<int16_t, T> &&
+                                                      !std::is_same_v<uint32_t, T> && !std::is_same_v<int32_t, T>,
+                                                  int> = 0>
+  T get_value() {
+    _receive_if_necessary(sizeof(T));
+    T network_value = 0;
+    std::copy_n(_start_position, sizeof(T), reinterpret_cast<char*>(&network_value));
+    _start_position += sizeof(T);
+    return network_value;
+  }
+
+  // String functions
+  std::string get_string(const size_t string_length, const bool has_null_terminator = true);
+  std::string get_string();
+
+  PostgresMessageType get_message_type();
+
+ private:
+  std::shared_ptr<SocketType> _socket;
+  void _receive_if_necessary(const size_t bytes_required = 1);
+};
+
+// Dedicated buffer for write operations. The ring buffer gets extended by methods for writing different data types.
+template <typename SocketType>
+class WriteBuffer : public RingBuffer {
+ public:
+  explicit WriteBuffer(const std::shared_ptr<SocketType> socket) : _socket(socket) {}
+
+  // Flush whole buffer, e. g. after a finished request
+  void flush(const size_t bytes_required = 0);
+
+  // Put string into the buffer. If the string is longer than the buffer itself the buffer will flush automatically.
+  void put_string(const std::string& value, const bool terminate = true);
+
+  // Put numerical values into the buffer. Values will be converted into the correct byte order if necessary.
+  template <typename T, typename std::enable_if_t<std::is_same_v<uint32_t, T> || std::is_same_v<int32_t, T>, int> = 0>
+  void put_value(const T network_value) {
```

Comment:
I wonder if this is easier to read and has less duplication if we work with `if constexpr` in a single implementation